pkg/imgutil/dockerconfigresolver/credentialsstore.go

```diff
@@ -0,0 +1,182 @@
+// if it is stored / to be stored in a different place (like an OS keychain, with docker credential helpers)
+func (cs *CredentialsStore) FileStorageLocation(registryURL *RegistryURL) string {
+	if store, isFile := (cs.dockerConfigFile.GetCredentialsStore(registryURL.CanonicalIdentifier())).(isFileStore); isFile {
+		return store.GetFilename()
+	}
+
+	return ""
+}
+
+// Retrieve gets existing credentials from the store for a certain registry.
+// If none are found, an empty Credentials struct is returned.
+// If we hard-fail reading from the store, indicative of a broken system, we wrap the error with ErrUnableToRetrieve
+func (cs *CredentialsStore) Retrieve(registryURL *RegistryURL, checkCredStore bool) (*Credentials, error) {
+	var err error
+	returnedCredentials := &Credentials{}
+
+	// As long as we depend on .ServerAddress, make sure it is populated correctly
+	// It does not matter what was stored - the docker cli clearly has issues with this
+	// What matters is that the credentials retrieved from the docker credentials store are *for that registryURL*
+	// and that is what ServerAddress should point to
+	defer func() {
+		if registryURL.Namespace != nil {
+			returnedCredentials.ServerAddress = fmt.Sprintf("%s%s?%s", registryURL.Host, registryURL.Path, registryURL.RawQuery)
+		} else {
+			returnedCredentials.ServerAddress = registryURL.Host
+		}
+	}()
+
+	if !checkCredStore {
+		return returnedCredentials, nil
+	}
+
+	log.L.Debugf("retrieving credentials from store for %s", registryURL.CanonicalIdentifier())
+
+	// Get the legacy variants (w/o scheme or port), and iterate over until we find one with credentials
+	variants := registryURL.AllIdentifiers()
+
+	log.L.Debugf("all identifiers %v", variants)
+
+	for _, identifier := range variants {
+		log.L.Debugf("looking up %s", identifier)
+		var credentials types.AuthConfig
+		// Note that Get does not raise an error on ENOENT
+		credentials, err = cs.dockerConfigFile.GetCredentialsStore(identifier).Get(identifier)
+		if err != nil {
+			continue
+		}
+		returnedCredentials = &credentials
+		// Clean-up the username
+		returnedCredentials.Username = strings.TrimSpace(returnedCredentials.Username)
+		// Stop here if we found credentials with this variant
+		if returnedCredentials.IdentityToken != "" ||
+			returnedCredentials.Username != "" ||
+			returnedCredentials.Password != "" ||
+			returnedCredentials.RegistryToken != "" {
+			log.L.Debugf("found credentials %v", returnedCredentials)
+			break
+		}
+	}
+
+	// (Last non nil) credential store error gets wrapped into ErrUnableToRetrieve
+	if err != nil {
+		err = errors.Join(ErrUnableToRetrieve, err)
+	}
+
+	return returnedCredentials, err
+}
+
+// isFileStore is an internal mock interface purely meant to help identify that the docker credential backend is a filesystem one
+type isFileStore interface {
+	IsFileStore() bool
+	GetFilename() string
+}
```

Comment:
interface should be moved to the begining of the file, or at least before the implementation.

Reply:
It is private and only used internally in this file - there is no implementation, as it is used solely to cast docker credentials store.

In that context, do you feel it should still be moved up? (no strong opinion on my side)